WordPress/src/main/java/org/wordpress/android/ui/posts/PostEditorAnalyticsSession.java

```diff
@@ -0,0 +1,132 @@
+            mBlogType = "jetpack";
+        } else {
+            mBlogType = "core";
+        }
+
+        // fill in mContentType
+        String postContent = post.getContent();
+        if (TextUtils.isEmpty(post.getContent())) {
+            mContentType = "new";
+        } else if (PostUtils.contentContainsGutenbergBlocks(postContent)) {
+            mContentType = "gutenberg";
+        } else {
+            mContentType = "classic";
+        }
+    }
+
+    public void start(boolean hasUnsupportedBlocks) {
+        if (!mStarted) {
+            mHasUnsupportedBlocks = hasUnsupportedBlocks;
+            Map<String, Object> properties = getCommonProperties();
+            AnalyticsTracker.track(Stat.EDITOR_SESSION_START, properties);
+            mStarted = true;
+        } else {
+            AppLog.w(T.EDITOR, "An editor session cannot be started more than once, "
+                               + "unless it's due to rotation or Editor switch");
+        }
+    }
+
+    public void switchEditor(Editor editor) {
+        mCurrentEditor = editor;
+        Map<String, Object> properties = getCommonProperties();
+        AnalyticsTracker.track(Stat.EDITOR_SESSION_SWITCH_EDITOR, properties);
+    }
+
+    public void setOutcome(Outcome newOutcome) {
+        // on WPiOS we're using `forceOutcome` and some exception logic to allow / not allow forcing an outcome,
+        // but in WPAndroid we slightly changed the interface as we're not
+        // really `forcing` but just `setting` the outcome on specific places where we know what happened
+        // or is going to happen.
+        // Session ending will only properly end the session with the outcome already being set.
+        mOutcome = newOutcome;
+    }
+
+    public void end() {
+        // don't try to send an "end" event if the session wasn't started in the first place
+        if (mStarted) {
+            if (mOutcome == null) {
+                // outcome should have already been set with setOutcome at specific user actions
+                // if outcome is still unknown, chances are Activity was killed / user cancelled so, set to CANCEL.
+                mOutcome = Outcome.CANCEL;
+            }
+            Map<String, Object> properties = getCommonProperties();
+            properties.put(KEY_OUTCOME, mOutcome.toString().toLowerCase(Locale.ROOT));
+            AnalyticsTracker.track(Stat.EDITOR_SESSION_END, properties);
+        }
+    }
+
+    private Map<String, Object> getCommonProperties() {
+        Map<String, Object> properties = new HashMap<>();
+        properties.put(KEY_EDITOR, mCurrentEditor.toString().toLowerCase(Locale.ROOT));
+        properties.put(KEY_CONTENT_TYPE, mContentType);
+        properties.put(KEY_POST_TYPE, mPostType);
+        properties.put(KEY_BLOG_TYPE, mBlogType);
+        properties.put(KEY_SESSION_ID, mSessionId);
+        properties.put(KEY_HAS_UNSUPPORTED_BLOCKS, mHasUnsupportedBlocks ? "1" : "0");
```

Comment:
Hmm, do we need to keep this property in the "common" set? This particular property is only accurate when the session starts. It might still hold true on session_end but as we don't re-calculate it we can't be sure. Can we remove it from the common?

Reply:
I think you're right - but this differs from the iOS implementation which includes `hasUnsupportedBlocks` in the `commonProperties` https://github.com/wordpress-mobile/WordPress-iOS/pull/11040/files#diff-d51d8e640f5dd1927766cb6c5136bf02R73 wdyt @koke @hypest ?

Reply:
I think it'd be accurate 99% of the time since you can only get unsupported after start if you mess with the HTML. In any case, I know the name says "has unsupported", but the spirit of the metric was to reflect if the post had unsupported blocks when it was initially opened.

We can revisit afterwards, but I'd leave it as is for consistency

Reply:
Ah, if iOS has it like this, 💯 OK to leave it as is to match the behavior.